### PR TITLE
[RLlib] In the Json_writer convert all non string keys to keys

### LINKED ISSUE #33789: [RLlib] something broken about the way that torch ppo writes out to json
### What happened + What you expected to happen

when using the output writer, torch ppo writes out sample batches in the incorrect format:

`“value”: {”type”: “MultiAgentBatch”...`

instead of just

`{”type”: “MultiAgentBatch”...`

### Versions / Dependencies

NA

### Reproduction script

Useful debug script:
https://gist.github.com/avnishn/c8599a9947edcd9934421713ed82f175
```
def generate_offline_data(path: str):
    print(f"Generating offline data for training at {path}")
    trainer = RLTrainer(
        algorithm="PPO",
        run_config=RunConfig(stop={"timesteps_total": 5000}),
        config={
            "env": "CartPole-v1",
            "output": "dataset",
            "output_config": {
                "format": "json",
                "path": path,
                "max_num_samples_per_file": 1,
            },
            "batch_mode": "complete_episodes",
            "framework": "tf"
        },
    )
    trainer.fit()
def train_rl_bc_offline(path: str, num_workers: int, use_gpu: bool = False) -> Result:
    print("Starting offline training")
    dataset = ray.data.read_json(
        path, parallelism=num_workers, ray_remote_args={"num_cpus": 1}
    )

    trainer = RLTrainer(
        run_config=RunConfig(stop={"training_iteration": 5}),
        scaling_config=ScalingConfig(num_workers=num_workers, use_gpu=use_gpu),
        datasets={"train": dataset},
        algorithm=BC,
        config={
            "env": "CartPole-v1",
            "framework": "tf",
            "evaluation_num_workers": 1,
            "evaluation_interval": 1,
            "evaluation_config": {"input": "sampler"},
            "framework": "tf"
        },
    )

    # Todo (krfricke/xwjiang): Enable checkpoint config in RunConfig
    # result = trainer.fit()
    tuner = Tuner(
        trainer,
        _tuner_kwargs={"checkpoint_at_end": True},
    )
    result = tuner.fit()[0]
    return result
ray.init(num_cpus=8)

path = "/tmp/out"
generate_offline_data(path)
result = train_rl_bc_offline(path=path, num_workers=2, use_gpu=False)
```
Should work with torch but doesn’t because of the way that sample batches are created.

### Issue Severity

None

### PR DESCRIPTION
**Pull request by avnishn**
Signed-off-by: Avnish <avnishnarayan@gmail.com>

Currently ray data has 2 formats in which it can store python dictionaries as datasets internally.

1. it will try to use pyarrow to store the underlying data.
2. it will try to use pandas / dictionaries to store the underlying data in the dataset.
In case 1, when writing data out to a json file, the data will be written out in the expected schema. e.g. {YOUR DATA: ..., }.
IN case 2, when writing data out to json, the schema is {"value": {YOUR DATA: ....}.

Ray data tries to go for 1 always but falls back onto 2 in the case that the dictionary being converted to a dataset can't be processed by arrow.

One of the reasons for fall back from 1 to 2 is if any of the keys in this dictionary are not strings

IN RLLIB this happens whenever our sample batch includes environment infos, and those environment infos have non string keys. This happens when the env infos are mapping agent ids to infos and those agent ids are integers.

Env infos are included as a part of the batch with torch but not tensorflow which is why we saw this issue when we made the default framework torch last week.

The solution for now is to convert non string keys to string keys, but to also force ray data to always do 1 and use py arrow to store the underlying data. We do this with the `output_arrow_format` flag added to `from_list` in  #33837

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datasource/binary_datasource.py
Comment (avnishn) [inline python/ray/data/datasource/binary_datasource.py:1]: all ray data changes in this pr are from #33837 and should be ignored during review.

File: doc/source/ray-air/examples/rl_offline_example.ipynb
Comment (avnishn) [inline doc/source/ray-air/examples/rl_offline_example.ipynb:1]: ptal at this file

File: rllib/offline/dataset_writer.py
Comment (avnishn) [inline rllib/offline/dataset_writer.py:1]: ptal at this file

File: rllib/offline/json_writer.py
Comment (avnishn) [inline rllib/offline/json_writer.py:1]: ptal at this file

File: rllib/offline/json_writer.py
Comment (gjoliver) [inline rllib/offline/json_writer.py:120]: I am actually not sure if you can do this.
like what should we do after we read this json data back in? how would we know which key to convert back into integers?
maybe just clear infos column here for now?

Comment (avnishn) [reply rllib/offline/json_writer.py:120]: Ok I'm in agreement sure.

**Comment by ollie-iterators** [comment]
Strange, I would think the error about "can't pickle _thread.lock objects" would be related to https://github.com/ray-project/ray/pull/33660, but it appears that "Dataset tests (Arrow nightly)" is passing right now.

**Comment by avnishn** [comment]
we need to wait for https://github.com/ray-project/ray/pull/33837 to be merged but otherwise this is ready to go afaict.

**Review by gjoliver — COMMENTED** [review]
ok, this looks reasonable.
can we wait for Amog to merge his PR, then rebase here, so we can have a final look before merge?
thanks for the fix.

**Review by gjoliver — APPROVED** [review]

**Comment by avnishn** [comment]
as it turns out there was a way to store infos in tensorflow for training.
Its not publicly documented, but there is 1 test for it, so there were 
2 occurences in the codebase. This whole pr pretty much disables you from being able 
to write env infos out in a sample batch, so we'll allow it to still be used for training,
but not allow it for output writing.

I removed the test where we check for env infos in the output batch for the time being.

File: rllib/tests/test_io.py
Comment (avnishn) [inline rllib/tests/test_io.py:83]: as it turns out there was a way to store infos in tensorflow for training.
Its not publicly documented, but there is 1 test for it, so there were
2 occurences in the codebase. This whole pr pretty much disables you from being able
to write env infos out in a sample batch, so we'll allow it to still be used for training,
but not allow it for output writing.

I removed the test where we check for env infos in the output batch for the time being.

cc @gjoliver

**Pull request merged**
